staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -678,43 +683,83 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
 	client := r.c.Client
 	if client == nil {
 		client = http.DefaultClient
 	}
-	r.backoff.Sleep(r.backoff.CalculateBackoff(r.URL()))
-	resp, err := client.Do(req)
-	updateURLMetrics(ctx, r, resp, err)
-	if r.c.base != nil {
+
+	// Throttle the first try before setting up the timeout configured on the
+	// client. We don't want a throttled client to return timeouts to callers
+	// before it makes a single request.
+	if err := r.tryThrottle(ctx); err != nil {
+		return nil, err
+	}
+
+	var retryAfter *RetryAfter
+	for {
+		url := r.URL().String()
+		req, err := http.NewRequest(r.verb, url, r.body)
 		if err != nil {
-			r.backoff.UpdateBackoff(r.c.base, err, 0)
-		} else {
-			r.backoff.UpdateBackoff(r.c.base, err, resp.StatusCode)
+			return nil, err
 		}
-	}
-	if err != nil {
-		// The watch stream mechanism handles many common partial data errors, so closed
-		// connections can be retried in many cases.
-		if net.IsProbableEOF(err) || net.IsTimeout(err) {
-			return watch.NewEmptyWatch(), nil
+		req = req.WithContext(ctx)
+		req.Header = r.headers
+
+		r.backoff.Sleep(r.backoff.CalculateBackoff(r.URL()))
+		if retryAfter != nil {
+			// We are retrying the request that we already send to apiserver
+			// at least once before.
+			// This request should also be throttled with the client-internal rate limiter.
+			if err := r.tryThrottleWithInfo(ctx, retryAfter.Reason); err != nil {
+				return nil, err
+			}
+			retryAfter = nil
 		}
-		return nil, err
-	}
-	if resp.StatusCode != http.StatusOK {
-		defer resp.Body.Close()
-		if result := r.transformResponse(resp, req); result.err != nil {
-			return nil, result.err
+
+		resp, err := client.Do(req)
+		updateURLMetrics(ctx, r, resp, err)
+		if r.c.base != nil {
+			if err != nil {
+				r.backoff.UpdateBackoff(r.c.base, err, 0)
+			} else {
+				r.backoff.UpdateBackoff(r.c.base, err, resp.StatusCode)
+			}
 		}
-		return nil, fmt.Errorf("for request %s, got status: %v", url, resp.StatusCode)
+		if err != nil {
+			// The watch stream mechanism handles many common partial data errors, so closed
+			// connections can be retried in many cases.
+			if net.IsProbableEOF(err) || net.IsTimeout(err) {
+				return watch.NewEmptyWatch(), nil
+			}
+			return nil, err
+		}
+		if resp.StatusCode == http.StatusOK {
+			return r.newStreamWatcher(resp)
+		}
+
+		var retry bool
+		retryAfter, retry = r.retry.NextRetry(resp, nil)
+		if !retry {
+			err = func() error {
+				defer resp.Body.Close()
+				if result := r.transformResponse(resp, req); result.err != nil {
+					return result.err
+				}
+				return fmt.Errorf("for request %s, got status: %v", url, resp.StatusCode)
+			}()
+			return nil, err
+		}
+
+		if err := r.retry.CleanBeforeRetry(resp, r.body); err != nil {
+			klog.V(4).Infof("Could not retry request - %v", err)
+			return nil, err
+		}
+		klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", retryAfter.Wait, retryAfter.Attempts, url)
+		r.backoff.Sleep(retryAfter.Wait)
```

Comment:
should have context passed through for an early return

Reply:
@deads2k  `r.backoff` currently has no room for `context`
```
type BackoffManager interface {
	UpdateBackoff(actualUrl *url.URL, err error, responseCode int)
	CalculateBackoff(actualUrl *url.URL) time.Duration
	Sleep(d time.Duration)
}
```
so, I need to extend this with `SleepWithContext(ctx context.Context, d time.Duration)`. This is a breaking change, we  allow users  to define their own `BackoffManager` implementation by `func (r *Request) BackOff(manager BackoffManager) *Request`.

Reply:
There might be a way to use `CalculateBackoff` to get the sleep duration and avoid the `r.backoff.Sleep` method which defeats the purpose of the interface and also will fail unit tests which rely on `Sleep` being called.